src/driver.rs

```diff
@@ -87,44 +112,84 @@ impl Driver {
     /// Runs the Driver event loop.
     pub async fn run(&mut self) -> Result<()> {
         info!("Starting driver event loop.");
-        while let Some(cmd) = self.cmd_rx.recv().await {
-            // TODO: these are kind of janky, we might want to move to a polling-based architecture
-            if let Some(Ok(res)) = poll_fn(|cx| {
-                Pin::new(&mut self.reader_join_handle)
-                    .as_mut()
-                    .poll_unpin(cx)
-            })
-            .now_or_never()
-            {
-                self.alert_tx.send(Alert::DriverError(eyre!(
-                    "Reader task exited unexpectedly."
-                )))?;
-                return res;
+        loop {
+            select! {
+                res = self.network.0.next() => {
+                    if let Some(res) = res {
+                        self.handle_network_event(res)?;
+                    }
+                },
+                cmd = self.cmd_rx.recv() => {
+                    if let Some(cmd) = cmd {
+                        self.handle_driver_command(cmd)?;
+                    } else {
+                        info!("Driver command channel closed.");
+                        break;
+                    }
+                }
             }
+        }
+        Ok(())
+    }
 
-            if let Some(Ok(res)) = poll_fn(|cx| {
-                Pin::new(&mut self.executor_join_handle)
-                    .as_mut()
-                    .poll_unpin(cx)
-            })
-            .now_or_never()
-            {
-                self.alert_tx.send(Alert::DriverError(eyre!(
-                    "Executor task exited unexpectedly."
-                )))?;
-                return res;
+    fn handle_network_event(&self, event: NetworkEvent) -> Result<()> {
+        match event {
+            NetworkEvent::NewListenAddr(addr) => {
+                info!("listening on {}", addr);
+            }
+            NetworkEvent::Message(msg) => {
+                debug!("received gossip message: {:?}", msg);
+                let block = SequencerBlock::from_bytes(&msg.data)?;
+                self.executor_tx
+                    .send(ExecutorCommand::BlockReceivedGossip {
+                        block: Box::new(block),
+                    })?;
             }
+            _ => debug!("received network event: {:?}", event),
+        }
 
-            match cmd {
-                DriverCommand::Shutdown => {
-                    self.shutdown()?;
-                    break;
-                }
-                DriverCommand::GetNewBlocks => {
-                    self.reader_tx
-                        .send(ReaderCommand::GetNewBlocks)
-                        .map_err(|e| eyre!("reader rx channel closed: {}", e))?;
-                }
+        Ok(())
+    }
+
+    fn handle_driver_command(&mut self, cmd: DriverCommand) -> Result<()> {
+        // TODO: these are kind of janky, we might want to move to a polling-based architecture
```

Comment:
the purpose of this was basically to get if the reader/executor tasks had returned early, meaning they had errored/panicked. I updated this so that the reader/executor JoinHandles are now handled in the main function in their own task